File: tools/check_workflow_topology.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "yaml"

ROOT = File.expand_path("..", __dir__)
WORKFLOW_ROOT = File.join(ROOT, ".github", "workflows")
RELEASE_PATH = File.join(WORKFLOW_ROOT, "release.yml")

def fail_check(message)
  warn("workflow topology: #{message}")
  exit(1)
end

def load_workflow(path)
  YAML.safe_load(File.read(path), aliases: true)
rescue Psych::Exception => error
  fail_check("#{path}: #{error.message}")
end

def triggers(workflow)
  # YAML 1.1 treats the unquoted GitHub key `on` as boolean true.
  workflow["on"] || workflow[true] || {}
end

def needs(job)
  Array(job["needs"])
end

workflows = Dir[File.join(WORKFLOW_ROOT, "*.yml")].sort.to_h do |path|
  [File.basename(path), load_workflow(path)]
end
release = workflows.fetch("release.yml")
release_jobs = release.fetch("jobs")

workflows.each do |name, workflow|
  contents = workflow.fetch("permissions").fetch("contents")
  fail_check("#{name} must default to contents: read") unless contents == "read"
end

publish_permissions = release_jobs.fetch("publish").fetch("permissions")
fail_check("only the publish job may request contents: write") unless
  publish_permissions.fetch("contents") == "write"

parallel_workflows = {
  "quality-gate" => "quality.yml",
  "core-gate" => "core.yml",
  "native-gate" => "native-integration.yml",
  "godot-gate" => "godot-compatibility.yml",
  "android-gate" => "android.yml",
  "web-gate" => "web.yml",
  "ios-gate" => "ios.yml",
  "host-components" => "host-components.yml",
}.freeze

parallel_workflows.each do |job_name, workflow_name|
  workflow_triggers = triggers(workflows.fetch(workflow_name))
  fail_check("#{workflow_name} must support workflow_call") unless
    workflow_triggers.key?("workflow_call")
  fail_check("#{workflow_name} must support standalone workflow_dispatch") unless
    workflow_triggers.key?("workflow_dispatch")
  call = workflow_triggers.fetch("workflow_call")
  fail_check("#{workflow_name} must require GDPP_TOKEN") unless
    call.fetch("secrets").fetch("GDPP_TOKEN").fetch("required")
  fail_check("#{workflow_name} must accept source_ref") unless
    call.fetch("inputs").key?("source_ref")

  job = release_jobs.fetch(job_name)
  fail_check("#{job_name} must invoke #{workflow_name}") unless
    job["uses"] == "./.github/workflows/#{workflow_name}"
  fail_check("#{job_name} must start immediately after preflight") unless
    needs(job) == ["preflight"]
end

package_workflow = workflows.fetch("package-release.yml")
package_triggers = triggers(package_workflow)
fail_check("package-release.yml must support workflow_call") unless
  package_triggers.key?("workflow_call")
fail_check("package-release.yml cannot run without producer artifacts") if
  package_triggers.key?("workflow_dispatch")

package_job = release_jobs.fetch("packages")
fail_check("packages must invoke package-release.yml") unless
  package_job["uses"] == "./.github/workflows/package-release.yml"
expected_package_needs = (parallel_workflows.keys + ["preflight"]).sort
fail_check("packages must wait for every test and component producer") unless
  needs(package_job).sort == expected_package_needs

package_smoke_workflow = workflows.fetch("release-package-smoke.yml")
package_smoke_triggers = triggers(package_smoke_workflow)
fail_check("release-package-smoke.yml must support workflow_call") unless
  package_smoke_triggers.key?("workflow_call")
fail_check("release-package-smoke.yml cannot run without assembled package artifacts") if
  package_smoke_triggers.key?("workflow_dispatch")
smoke_matrix = package_smoke_workflow.fetch("jobs").fetch("desktop")
  .fetch("strategy").fetch("matrix").fetch("include")
smoke_pairs = smoke_matrix.map do |entry|
  [entry.fetch("archive"), entry.fetch("os")]
end.sort
expected_smoke_pairs = [
  ["gdpp.zip", "macos-15"],
  ["gdpp.zip", "ubuntu-22.04"],
  ["gdpp.zip", "windows-2025"],
  ["gdpp-lite.zip", "macos-15"],
  ["gdpp-lite.zip", "windows-2025"],
].sort
fail_check("installed package smoke matrix is incomplete") unless
  smoke_pairs == expected_smoke_pairs

package_smoke_job = release_jobs.fetch("package-smoke")
fail_check("package-smoke must invoke release-package-smoke.yml") unless
  package_smoke_job["uses"] == "./.github/workflows/release-package-smoke.yml"
fail_check("package-smoke must retain source identity and wait for packages") unless
  needs(package_smoke_job).sort == ["packages", "preflight"]

readiness_job = release_jobs.fetch("readiness")
expected_readiness_needs = (parallel_workflows.keys + ["packages", "package-smoke"]).sort
fail_check("readiness must aggregate every delivery result") unless
  needs(readiness_job).sort == expected_readiness_needs
fail_check("publish must retain source identity and depend on readiness") unless
  needs(release_jobs.fetch("publish")).sort == ["preflight", "readiness"]

pull_request_workflows = workflows.each_with_object([]) do |(name, workflow), names|
  names << name if triggers(workflow).key?("pull_request")
end
fail_check("public workflows must never execute for pull requests") unless
  pull_request_workflows.empty?

workflows.each do |name, workflow|
  next if name == "release.yml"

  source_checkouts = workflow.fetch("jobs").values.sum do |job|
    Array(job["steps"]).count { |step| step["uses"] == "./.github/actions/checkout-source" }
  end
  fail_check("#{name} must check out private source") if source_checkouts.zero?
end

puts(
  "Validated #{parallel_workflows.length} parallel producers, " \
  "one gated package stage, installed macOS/Linux/Windows package smokes, and one publish stage.",
)

File: tools/check_workflow_topology.py
#!/usr/bin/env python3
"""Validate the public CI control plane and immutable private-source contract."""

from __future__ import annotations

from pathlib import Path
from typing import Any

import yaml


ROOT = Path(__file__).resolve().parents[1]
WORKFLOW_ROOT = ROOT / ".github/workflows"


def fail(message: str) -> None:
    raise SystemExit(f"workflow topology: {message}")


def load_workflow(path: Path) -> dict[str, Any]:
    try:
        workflow = yaml.safe_load(path.read_text(encoding="utf-8"))
    except yaml.YAMLError as error:
        fail(f"{path}: {error}")
    if not isinstance(workflow, dict):
        fail(f"{path}: workflow root must be a mapping")
    return workflow


def triggers(workflow: dict[str, Any]) -> dict[str, Any]:
    value = workflow.get("on", workflow.get(True, {}))
    return value if isinstance(value, dict) else {}


def needs(job: dict[str, Any]) -> list[str]:
    value = job.get("needs", [])
    return value if isinstance(value, list) else [value]


def require_private_source_contract(name: str, workflow: dict[str, Any]) -> None:
    call = triggers(workflow).get("workflow_call")
    if not isinstance(call, dict):
        fail(f"{name} must support workflow_call")
    inputs = call.get("inputs", {})
    secrets = call.get("secrets", {})
    if "source_ref" not in inputs:
        fail(f"{name} must accept source_ref")
    if not secrets.get("GDPP_TOKEN", {}).get("required"):
        fail(f"{name} must require GDPP_TOKEN")


def main() -> int:
    workflows = {
        path.name: load_workflow(path)
        for path in sorted(WORKFLOW_ROOT.glob("*.yml"))
    }
    release = workflows["release.yml"]
    release_jobs = release["jobs"]

    for name, workflow in workflows.items():
        if workflow.get("permissions", {}).get("contents") != "read":
            fail(f"{name} must default to contents: read")
        if "pull_request" in triggers(workflow):
            fail(f"{name} must never execute for pull requests")

    publish = release_jobs["publish"]
    if publish.get("permissions", {}).get("contents") != "write":
        fail("only the publish job may request contents: write")
    if set(triggers(release)) != {"workflow_dispatch"}:
        fail("release.yml must expose only workflow_dispatch")

    parallel = {
        "quality-gate": "quality.yml",
        "core-gate": "core.yml",
        "native-gate": "native-integration.yml",
        "godot-gate": "godot-compatibility.yml",
        "android-gate": "android.yml",
        "web-gate": "web.yml",
        "ios-gate": "ios.yml",
        "host-components": "host-components.yml",
    }
    for job_name, workflow_name in parallel.items():
        workflow = workflows[workflow_name]
        require_private_source_contract(workflow_name, workflow)
        if "workflow_dispatch" not in triggers(workflow):
            fail(f"{workflow_name} must support standalone workflow_dispatch")
        job = release_jobs[job_name]
        if job.get("uses") != f"./.github/workflows/{workflow_name}":
            fail(f"{job_name} must invoke {workflow_name}")
        if needs(job) != ["preflight"]:
            fail(f"{job_name} must start immediately after preflight")
        if job.get("with", {}).get("source_ref") != (
            "${{ needs.preflight.outputs.source_sha }}"
        ):
            fail(f"{job_name} must use the immutable preflight source SHA")
        if job.get("secrets", {}).get("GDPP_TOKEN") != "${{ secrets.GDPP_TOKEN }}":
            fail(f"{job_name} must pass only the explicit source token")

    package_workflow = workflows["package-release.yml"]
    require_private_source_contract("package-release.yml", package_workflow)
    if "workflow_dispatch" in triggers(package_workflow):
        fail("package-release.yml cannot run without producer artifacts")
    package_job = release_jobs["packages"]
    if package_job.get("uses") != "./.github/workflows/package-release.yml":
        fail("packages must invoke package-release.yml")
    expected_package_needs = sorted([*parallel, "preflight"])
    if sorted(needs(package_job)) != expected_package_needs:
        fail("packages must wait for preflight and every producer")

    smoke_workflow = workflows["release-package-smoke.yml"]
    require_private_source_contract("release-package-smoke.yml", smoke_workflow)
    smoke_dispatch = triggers(smoke_workflow).get("workflow_dispatch", {})
    smoke_dispatch_inputs = smoke_dispatch.get("inputs", {})
    if not smoke_dispatch_inputs.get("source_ref", {}).get("required"):
        fail("release-package-smoke.yml dispatch must require an immutable source_ref")
    if not smoke_dispatch_inputs.get("artifact_run_id", {}).get("required"):
        fail("release-package-smoke.yml dispatch must require an assembled artifact run")
    smoke_matrix = smoke_workflow["jobs"]["desktop"]["strategy"]["matrix"]["include"]
    actual_smokes = sorted((entry["archive"], entry["os"]) for entry in smoke_matrix)
    expected_smokes = sorted(
        (
            ("gdpp.zip", "macos-15"),
            ("gdpp.zip", "ubuntu-22.04"),
            ("gdpp.zip", "windows-2025"),
            ("gdpp-lite.zip", "macos-15"),
            ("gdpp-lite.zip", "windows-2025"),
        )
    )
    if actual_smokes != expected_smokes:
        fail("installed package smoke matrix is incomplete")
    smoke_job = release_jobs["package-smoke"]
    if smoke_job.get("uses") != "./.github/workflows/release-package-smoke.yml":
        fail("package-smoke must invoke release-package-smoke.yml")
    if sorted(needs(smoke_job)) != ["packages", "preflight"]:
        fail("package-smoke must retain source identity and wait for packages")

    readiness_needs = sorted([*parallel, "packages", "package-smoke"])
    if sorted(needs(release_jobs["readiness"])) != readiness_needs:
        fail("readiness must aggregate every delivery result")
    if sorted(needs(publish)) != ["preflight", "readiness"]:
        fail("publish must retain source identity and depend on readiness")

    for name, workflow in workflows.items():
        if name == "release.yml":
            continue
        source_checkouts = sum(
            step.get("uses") == "./.github/actions/checkout-source"
            for job in workflow["jobs"].values()
            for step in job.get("steps", [])
        )
        if source_checkouts == 0:
            fail(f"{name} must check out private source")

    print(
        f"Validated {len(parallel)} parallel producers, one gated package stage, "
        "installed macOS/Linux/Windows package smokes, and one publish stage."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
